contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,826 @@
+}
+```
+
+Additionally we introduce the following constant.
+
+```go
+// StatefulSetRevisionLabel is the label used by StatefulSet controller to track
+// which version of StatefulSet's StatefulSetSpec was used generate a Pod.
+const StatefulSetRevisionLabel = "statefulset.kubernetes.io/revision"
+
+```
+## StatefulSet Controller
+The StatefulSet controller will watch for modifications to StatefulSet and Pod 
+API objects. When a StatefulSet is created or updated, or when one 
+of the Pods in a StatefulSet is updated or deleted, the StatefulSet
+controller will attempt to create, update, or delete Pods to conform the 
+current state of the system to the user declared [target state](#target-state). 
+
+### Revised Controller Algorithm
+The StatefulSet controller will use the following algorithm to continue to 
+make progress toward the user declared [target state](#target-state) while 
+respecting the controller's 
+[identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+[deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+guarantees. The StatefulSet controller will use the technique proposed in 
+[Controller History](https://github.com/kubernetes/community/pull/594) to 
+snapshot and version its [target Object state](#target-pod-state).
+
+1. The controller will reconstruct the 
+[revision history](#history-reconstruction) of the StatefulSet.
+1. The controller will 
+[process any updates to its StatefulSetSpec](#specification-updates) to 
+ensure that the StatefulSet's revision history is consistent with the user 
+declared desired state.
+1. The controller will select all Pods in the StatefulSet, filter any Pods not 
+owned by the StatefulSet, and sort the remaining Pods in ordinal order.
+1. For all created Pods, the controller will perform any nessecary 
+[non-destructive state reconciliation](#pod-state-reconciliation).
+1. If any Pods with ordinals in the sequence `[0,.Spec.Replicas)` have not been 
+created, for the Pod corresponding to the lowest such ordinal, the controller 
+will create the Pod with declared [target Pod state](#target-pod-state).
+1. If all Pods in the sequence `[0,.Spec.Replicas)` have been created, but if any 
+do not have a Ready Condition, the StatefulSet controller will wait for these 
+Pods to either become Ready, or to be completely deleted. 
+1. If all Pods in the sequence `[0,.Spec.Replicas)` have a Ready Condition, and 
+if `.Spec.Replicas` is less than `.Status.Replicas`, the controller will delete 
+the Pod corresponding to the largest ordinal. This implies that scaling takes 
+precedence over Pod updates.
+1. If all Pods in the range `[0,.Spec.Replicas)` have a Status of Running and 
+a Ready Condition, if `.Spec.Replicas` is equal to `.Status.Replicas`, and if 
+there are Pods that do not match their [target Pod state](#target-pod-state), 
```

Comment:
@kargakis In the initial state above, the update to `A` has completed -- according to the [Update Completion section](https://github.com/kow3ns/community/blob/8caee797f7aad9380eeef2bf369b290af169db84/contributors/design-proposals/statefulset-update.md#update-completion), para [5.iii] indicates that after update completion, the controller sets `.Status.UpdateReplicas` to 0.

At any rate, when `.Status.UpdateRevision` is changed to `B`, `.Status.UpdatedReplicas` should be set to 0 since no pods are currently at revision `B` -- so the question still stands.